src/java/org/apache/cassandra/index/sai/disk/hnsw/CassandraOnHeapHnsw.java

```diff
@@ -294,6 +294,35 @@ public SegmentMetadata.ComponentMetadataMap writeData(IndexDescriptor indexDescr
                                                                                   postingsMap.keySet().size(), vectorValues.size());
         logger.debug("Writing graph with {} rows and {} distinct vectors", postingsMap.values().stream().mapToInt(VectorPostings::size).sum(), vectorValues.size());
 
+        final Map<Integer, Integer> newOrdinals = new HashMap<>();
+        boolean doRemaping = !postingsMap.values().stream().anyMatch(x -> x. getPostings().size() > 1);
```

Comment:
So this is checking that each vector in `postingsMap` has <= 1 key/row associated to it. What if we deleted a row for a vector, and now it's postings list is empty? Then we are not in the happy path situation Jonathan mentioned, so at the very least, we need to do this lower after accruing deleted ordinals

Also, in general I believe what we are looking for as the criteria for being able to re-map ordinals:
- we have n ordinals/vectors that are all live (i.e. nothing deleted). These get consecutive values 0, ..., n - 1
- we have also n consecutive row ids that map 1-to-1 with the ordinals (i.e. so ur check that all `VectorPostings` have size precisely 1 + all the row id's are consecutive)

If we have this setup, then row ids and ordinals are effectively one and the same thing (since each row id has a unique ordinal/vector in the HNSW, and vice versa each ordinal/vector corresponds to a unique row id in the SSTable), and we can permute.